// @adpt/cloud/dist/src/index.js
"use strict";
/*
 * Copyright 2018-2019 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./Compute"), exports);
tslib_1.__exportStar(require("./ConnectTo"), exports);
tslib_1.__exportStar(require("./Container"), exports);
tslib_1.__exportStar(require("./DockerHost"), exports);
tslib_1.__exportStar(require("./LocalCompute"), exports);
tslib_1.__exportStar(require("./LocalContainer"), exports);
tslib_1.__exportStar(require("./LocalDockerHost"), exports);
tslib_1.__exportStar(require("./NetworkService"), exports);
tslib_1.__exportStar(require("./Service"), exports);
const action = tslib_1.__importStar(require("./action"));
exports.action = action;
const aws = tslib_1.__importStar(require("./aws"));
exports.aws = aws;
const common_1 = require("./common");
exports.makeResourceName = common_1.makeResourceName;
const docker = tslib_1.__importStar(require("./docker"));
exports.docker = docker;
const gcloud = tslib_1.__importStar(require("./gcloud"));
exports.gcloud = gcloud;
const http = tslib_1.__importStar(require("./http"));
exports.http = http;
const k8s = tslib_1.__importStar(require("./k8s"));
exports.k8s = k8s;
const mongodb = tslib_1.__importStar(require("./mongodb"));
exports.mongodb = mongodb;
const nginx = tslib_1.__importStar(require("./nginx"));
exports.nginx = nginx;
const nodejs = tslib_1.__importStar(require("./nodejs"));
exports.nodejs = nodejs;
const postgres = tslib_1.__importStar(require("./postgres"));
exports.postgres = postgres;
const redis = tslib_1.__importStar(require("./redis"));
exports.redis = redis;
tslib_1.__exportStar(require("./handles"), exports);
tslib_1.__exportStar(require("./env"), exports);
