// @adpt/utils/dist/src/object_set.js
"use strict";
/*
 * Copyright 2018-2019 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const Set = require("@usys/collections-ts/set");
const json_stable_stringify_1 = tslib_1.__importDefault(require("json-stable-stringify"));
function equals(a, b) {
    return json_stable_stringify_1.default(a) === json_stable_stringify_1.default(b);
}
function hash(obj) {
    return json_stable_stringify_1.default(obj);
}
class ObjectSet {
    constructor(values, defaults) {
        this.defaults = defaults;
        this.store_ = new Set(undefined, equals, hash);
        if (values != null) {
            // @ts-ignore
            values.forEach((val) => this.store_.add(this.withDefaults(val)));
        }
    }
    get length() {
        return this.store_.length;
    }
    add(value) {
        return this.store_.add(this.withDefaults(value));
    }
    has(value) {
        return this.store_.has(this.withDefaults(value));
    }
    clear() {
        this.store_.clear();
    }
    delete(value) {
        return this.store_.delete(this.withDefaults(value));
    }
    forEach(callback, thisp) {
        this.store_.forEach((value, key) => {
            callback.call(thisp, value, key, this);
        });
    }
    withDefaults(value) {
        // @ts-ignore
        return this.defaults ? Object.assign({}, this.defaults, value) : value;
    }
    [Symbol.iterator]() {
        return this.store_.iterator();
    }
}
exports.ObjectSet = ObjectSet;
