// @adpt/core/dist/src/server/history.js
"use strict";
/*
 * Copyright 2018-2019 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var HistoryStatus;
(function (HistoryStatus) {
    HistoryStatus["preAct"] = "preAct";
    HistoryStatus["success"] = "success";
    HistoryStatus["failed"] = "failed";
    HistoryStatus["complete"] = "complete";
})(HistoryStatus = exports.HistoryStatus || (exports.HistoryStatus = {}));
function isHistoryStatus(val) {
    switch (val) {
        case "preAct":
        case "success":
        case "failed":
        case "complete":
            return true;
    }
    return false;
}
exports.isHistoryStatus = isHistoryStatus;
function isStatusComplete(status) {
    return status === HistoryStatus.success || status === HistoryStatus.failed;
}
exports.isStatusComplete = isStatusComplete;
