// @adpt/core/dist/src/exports.js
"use strict";
/*
 * Copyright 2018-2019 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
var hooks_1 = require("./hooks");
exports.callFirstInstanceWithMethod = hooks_1.callFirstInstanceWithMethod;
exports.callInstanceMethod = hooks_1.callInstanceMethod;
exports.callNextInstanceMethod = hooks_1.callNextInstanceMethod;
exports.callNextInstanceWithMethod = hooks_1.callNextInstanceWithMethod;
exports.hasInstanceMethod = hooks_1.hasInstanceMethod;
exports.useAsync = hooks_1.useAsync;
exports.useBuildHelpers = hooks_1.useBuildHelpers;
exports.useDependsOn = hooks_1.useDependsOn;
exports.useDeployedWhen = hooks_1.useDeployedWhen;
exports.useImperativeMethods = hooks_1.useImperativeMethods;
exports.useInstanceValue = hooks_1.useInstanceValue;
exports.useMethod = hooks_1.useMethod;
exports.useMethodFrom = hooks_1.useMethodFrom;
exports.useState = hooks_1.useState;
var jsx_1 = require("./jsx");
exports.childrenToArray = jsx_1.childrenToArray;
exports.cloneElement = jsx_1.cloneElement;
exports.createElement = jsx_1.createElement;
exports.Component = jsx_1.Component;
exports.DeferredComponent = jsx_1.DeferredComponent;
exports.PrimitiveComponent = jsx_1.PrimitiveComponent;
exports.defaultDeployedWhen = jsx_1.defaultDeployedWhen;
exports.isApplyStyle = jsx_1.isApplyStyle;
exports.isFinalDomElement = jsx_1.isFinalDomElement;
exports.isElement = jsx_1.isElement;
exports.isPartialFinalDomElement = jsx_1.isPartialFinalDomElement;
exports.isMountedElement = jsx_1.isMountedElement;
exports.isMountedPrimitiveElement = jsx_1.isMountedPrimitiveElement;
exports.isDeferredElement = jsx_1.isDeferredElement;
exports.isPrimitiveElement = jsx_1.isPrimitiveElement;
tslib_1.__exportStar(require("./builtin_components"), exports);
tslib_1.__exportStar(require("./deploy"), exports);
var dom_1 = require("./dom");
exports.build = dom_1.build;
exports.buildOnce = dom_1.buildOnce;
var css_1 = require("./css");
exports.concatStyles = css_1.concatStyles;
exports.Style = css_1.Style;
exports.rule = css_1.rule;
exports.Rule = css_1.Rule;
exports.ruleNoRematch = css_1.ruleNoRematch;
exports.findElementsInDom = css_1.findElementsInDom;
exports.findPathsInDom = css_1.findPathsInDom;
var dom_serialize_1 = require("./dom_serialize");
exports.serializeDom = dom_serialize_1.serializeDom;
tslib_1.__exportStar(require("./dom_build_data_recorder"), exports);
tslib_1.__exportStar(require("./dom_utils"), exports);
var error_1 = require("./error");
exports.BuildNotImplemented = error_1.BuildNotImplemented;
exports.ProjectBuildError = error_1.ProjectBuildError;
exports.ProjectCompileError = error_1.ProjectCompileError;
exports.ProjectRunError = error_1.ProjectRunError;
var handle_1 = require("./handle");
exports.handle = handle_1.handle;
exports.isHandle = handle_1.isHandle;
var keys_1 = require("./keys");
exports.isDefaultKey = keys_1.isDefaultKey;
var context_1 = require("./context");
exports.createContext = context_1.createContext;
exports.useContext = context_1.useContext;
var stack_1 = require("./stack");
exports.stack = stack_1.stack;
var state_1 = require("./state");
exports.createStateStore = state_1.createStateStore;
tslib_1.__exportStar(require("./ops"), exports);
var observers_1 = require("./observers");
exports.registerObserver = observers_1.registerObserver;
exports.gql = observers_1.gql;
exports.Observer = observers_1.Observer;
exports.ObserverNeedsData = observers_1.ObserverNeedsData;
exports.throwObserverErrors = observers_1.throwObserverErrors;
var status_1 = require("./status");
exports.defaultChildStatus = status_1.defaultChildStatus;
exports.errorToNoStatus = status_1.errorToNoStatus;
exports.gqlGetOriginalErrors = status_1.gqlGetOriginalErrors;
exports.mergeDefaultChildStatus = status_1.mergeDefaultChildStatus;
exports.noStatusOnError = status_1.noStatusOnError;
exports.noTransform = status_1.noTransform;
var dom_filter_1 = require("./utils/dom-filter");
exports.deepFilterElemsToPublic = dom_filter_1.deepFilterElemsToPublic;
const internal = tslib_1.__importStar(require("./internal"));
exports.internal = internal;
