// @adpt/cloud/dist/src/aws/observer_common.js
"use strict";
/*
 * Copyright 2019 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const json_stable_stringify_1 = tslib_1.__importDefault(require("json-stable-stringify"));
exports.infoSym = Symbol("adaptObserverInfo");
exports.withParamsProp = "withCredentials";
function computeQueryId(queryParams, fieldName, args) {
    return json_stable_stringify_1.default({
        awsRegion: queryParams.awsRegion,
        awsAccessKeyId: queryParams.awsAccessKeyId,
        fieldName,
        args,
    });
}
exports.computeQueryId = computeQueryId;
function opName(fieldName) {
    return fieldName[0].toLowerCase() + fieldName.substr(1);
}
exports.opName = opName;
