// @adpt/dom-parser/dist/src/dom.js
"use strict";
/*
 * Copyright 2018 Unbounded Systems, LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const ld = tslib_1.__importStar(require("lodash"));
function isDOMNode(n) {
    return n instanceof DOMNode;
}
exports.isDOMNode = isDOMNode;
class DOMNode {
    constructor(componentType, props, lifecycleInfo, uri, children) {
        this.componentType = componentType;
        this.lifecycleInfo = lifecycleInfo;
        this.uri = uri;
        this.parent = null;
        this.props = ld.clone(props);
        if (children != null) {
            if (children.length !== 0) {
                for (const child of children) {
                    if (isDOMNode(child)) {
                        child.parent = this;
                    }
                }
                this.props.children = (children.length === 1) ?
                    children[0] :
                    children;
            }
        }
    }
}
exports.DOMNode = DOMNode;
class DOMObject {
    constructor(uri, data) {
        this.uri = uri;
        this.data = data;
    }
}
exports.DOMObject = DOMObject;
function isDOMObject(n) {
    return n instanceof DOMObject;
}
exports.isDOMObject = isDOMObject;
